rfc-175-frontend-fewer-apps.md

```diff
@@ -0,0 +1,254 @@
+
+### Testing standardisation
+We propose standardising around RSpec as the test suite for the frontend, ideally [modernising the tests], for example by switching to request tests rather than controller tests and system tests rather than feature tests. RSpec is the [preferred testing suite] on GOV.UK and is favoured by many developers because the tests and testing output are easier to read. Of the 4 applications only [government-frontend] uses minitest at this point, so its test suite could be converted route-by-route as it is merged into [frontend]
+
+### Code structure
+The code structure will be standardised. As well as using the [default Rails directory structure], we propose the following standard for additional classes (e.g. services, api calls). The standard expands on the [conventions] documented in the developer docs.
+
+The host app will follow the default MVC structure of Rails applications, with some additional restrictions on what goes where that are frontend specific (listed below)
+
+```
+├── app
+│   ├── assets
+│   ├── controllers
+│   ├── helpers
+│   ├── models
+│   ├── presenters
+│   ├── views
+├── config
+│   ├── data
+│   ├── locales
+├── lib
+│   ├── tasks
+└── .gitignore
+```
+
+Under `/app`:
+
+`helpers`
+- General utility methods for a view.
+
+`models`
+- Frontend apps do not have database records, so models here are created to store data received from api calls (eg content items from content store), process data that will be written to apis (eg feedback items to zendesk), and information ingested from YAML/JSON/CSVs.
+
+`presenters`
+- Presentation related transformations on model data. It is a wrapper around a model for use in a view.
+
+Under `/lib`
+
+`tasks`
+- Rake tasks
+
+Use `lib` rather than `app/lib` for custom code that enhances the app but doesn’t fit into a directory in `app`
+
+Under `/config`
+
+`data`
+- YAML, JSON, CSV config files. Data files should be placed here rather than under `/lib/data`.
+
+`locales`
+- Static content like lists of contact information should ideally be stored in the locale files to promote localisation.
+
+Namespaces will be used to group functionality e.g. for StepBySteps, and resources in the routes file. This structure will be documented in README of the host app.
+Each application will be analysed before being consolidated to ensure redundant code is not being moved over.
+
+### How the proposal will be carried out
+This is obviously a very large piece of work that will take some time to complete. We have identified the following steps that will be needed.
+
+* Prepare `frontend` to be the host
```

Comment:
Not sure if this is the right RFC or https://github.com/alphagov/govuk-rfcs/pull/176 but will this involve frontend taking over static responsibilities like error pages and common assets like favicon?

Reply:
For the moment I think that will be part of the work in RFC-176, but yes - current suggestion is that this will go in the static content gem, but it might be that during the implementation of that that we discover it makes more sense for some of those to got into Frontend. I think I'd like to keep that open (but I will mention it in RFC-176, which I'm just updating now). Until we can work up some proofs of concept, I'm reluctant to commit to moving them from one single app to another single app (just because there's outstanding tech debt related to nginx clobbering HTTP responses from the frontend apps, and part of solving that is making sure that some error responses are devolved to individual apps), but there will be a decision gate where we work out where it ends up.

Reply:
Thanks for bringing this up!

Reply:
Ok cool. I'm not quite sure I understand how the gem makes it easier because I would have thought that, without static, we don't want every app to be able to render things like favicon etc and just want one nominated one - since only one app should do www.gov.uk/favicon.ico for instance, although error pages might be different - but very happy to move that discussion away from this RFC since it'll be covered in the static one.